Src/ILGPU/Context.cs

```diff
@@ -384,6 +385,70 @@ public DeviceCollection<TDevice> GetDevices<TDevice>()
                 : new DeviceCollection<TDevice>(new List<Device>());
         }
 
+        /// <summary>
+        /// Attempts to return the most optimal single device.
+        /// </summary>
+        /// <param name="preferCPU">Always returns CPU device 0.</param>
+        /// <returns>Selected device.</returns>
+        public Device GetPreferredDevice(bool preferCPU)
+        {
+            if (preferCPU)
+            {
+                return GetDevice<CPUDevice>(0);
+            }
+
+            var sorted = Devices
+                .OrderByDescending(d => d.MemorySize)
+                .Where(d => d.AcceleratorType != AcceleratorType.CPU);
+
+            return sorted.FirstOrDefault() ?? GetDevice<CPUDevice>(0);
+        }
+
+        /// <summary>
+        /// Attempts to return the most optimal set of devices.
+        /// </summary>
+        /// <param name="preferCPU">Always returns first CPU device.</param>
+        /// <param name="matchingDevicesOnly">Only returns matching devices.</param>
+        /// <returns>Selected devices.</returns>
+        public IEnumerable<Device> GetPreferredDevices(bool preferCPU, bool matchingDevicesOnly)
+        {
+            if (preferCPU)
+            {
+                return deviceMapping.TryGetValue(AcceleratorType.CPU, out var devices)
```

Comment:
This looks like `GetDevice<CPUDevice>(0)` again.

Reply:
This returns the entire list, GetDevice<CPUDevice>(0) only returns 1.

Reply:
ok, so maybe use `GetCPUDevices()` then.

Reply:
Is there a consensus on this?

Reply:
The PR is correct - `GetCPUDevices` returns `DeviceCollection`, which does not implement `IEnumerable<>T`.